src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -2,35 +2,200 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
 using System.Threading;
 
 namespace Microsoft.Win32.SafeHandles
 {
-    public sealed class SafeFileHandle : SafeHandleZeroOrMinusOneIsInvalid
+    public sealed partial class SafeFileHandle : SafeHandleZeroOrMinusOneIsInvalid
     {
-        private bool? _isAsync;
+        internal const FileOptions NoBuffering = (FileOptions)(0x20000000);
+        private FileOptions _fileOptions = (FileOptions)(-1);
+        private int _fileType = -1;
 
         public SafeFileHandle() : base(true)
         {
-            _isAsync = null;
         }
 
         public SafeFileHandle(IntPtr preexistingHandle, bool ownsHandle) : base(ownsHandle)
         {
             SetHandle(preexistingHandle);
-
-            _isAsync = null;
         }
 
-        internal bool? IsAsync
+        private SafeFileHandle(IntPtr preexistingHandle, bool ownsHandle, FileOptions fileOptions, int fileType) : base(ownsHandle)
         {
-            get => _isAsync;
-            set => _isAsync = value;
+            SetHandle(preexistingHandle);
+
+            _fileOptions = fileOptions;
+            _fileType = fileType;
         }
 
+        public bool IsAsync => (GetFileOptions() & FileOptions.Asynchronous) != 0;
+
+        internal bool CanSeek => !IsClosed && GetFileType() == Interop.Kernel32.FileTypes.FILE_TYPE_DISK;
+
+        internal bool IsPipe => GetFileType() == Interop.Kernel32.FileTypes.FILE_TYPE_PIPE;
+
         internal ThreadPoolBoundHandle? ThreadPoolBinding { get; set; }
 
-        protected override bool ReleaseHandle() =>
-            Interop.Kernel32.CloseHandle(handle);
+        internal static unsafe SafeFileHandle Open(string fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize)
+        {
+            using (DisableMediaInsertionPrompt.Create())
+            {
+                SafeFileHandle fileHandle = new SafeFileHandle(
+                    NtCreateFile(fullPath, mode, access, share, options, preallocationSize),
```

Comment:
> NtCreateFile guaranteed to fully support every path that CreateFileW supported? I ask because CreateFileW is a higher-level API, and I could certainly see it special-casing various inputs.

@JeremyKuhne what would be the best way to answer this quesiton? Read `CreateFileW` source code?

Reply:
@adamsitnik, @stephentoub 

We do have the luxury of seeing the Windows sources so we can certainly validate our presumptions, but we should temper that by real-world expectations of change. In this case you're _very_ unlikely to see things change out from under you as drivers have no choice but to call the `Nt` (actually the `Zw` entry point) APIs and numerous others are calling the documented `NtCreateFile`.

All paths go through central canonicalization APIs. This was a fundamental fact I leveraged when trying to convince Windows to open up long paths for Win32 paths. Everything effectively calls `GetFullPathName` then when passed to `Nt` APIs set up the prefix to `\??\` by either prepending it, changing `\\.\` to it, changing `\\?\` to it, or changing any other `\\` path to `\??\UNC\`. As long as you munge correctly you should be fine.

[Project Zero](https://googleprojectzero.blogspot.com/2016/02/the-definitive-guide-on-win32-to-nt.html) describes the process in some depth. While we can't use undocumented APIs in shipping code we have used them in tests. I believe another option to test behavior is to actually work with real files, using a local UNC path, etc.  You can use [`GetFinalPathNameByHandleW`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfinalpathnamebyhandlew) to get the NT version of the path. I'm not entirely sure what local UNCs will get you, I've fiddled with this API in [WInterop](https://github.com/jeremyKuhne/winterop), you might want to play with it more there.

I'll note that it is always a good idea to test the volume root directory behavior (e.g. `C:\`).

It might also be worth exploring just ensuring `\\?\` (as you would with `\??\`) and calling `CreateFileW`. Some of our pain is simply that the same path is getting normalized multiple times. `\\?\` avoids that.

One last thing I'll call out is that getting comfortable with using `NtCreateFile` opens up additional functionality. We're already using it for enumeration, one can imagine where this would give some speed with things such as new temporary directory functionality. There is also other things that might end up interesting such as opening files by ID.

Reply:
I am marking this as resolved based on offline conversation.